rfcs/proposed/parallel_block_for_task_arena/README.md

```diff
@@ -0,0 +1,189 @@
+
+<img src="completely_disable_new_behavior.png" width=800>
+
+There will be a question that we need to answer:
+* Do we see any value if arena potentially can transition from one to another state?
+
+To answer this question, the following scenarios should be considired:
+* What if different types of workloads are mixed in one application?
+* Different types of arenas can be used for different types of workloads.
+
+### When threads should leave?
+
+oneTBB itself can only guess when the ideal time to release threads from the arena is.
+Therefore, it does the best effort to preserve and enhance performance without completely
+messing composability guarantees (that is how delayed leave is implemented).
+
+As we already discussed, there are cases where it does not work perfectly,
+therefore customers that want to further optimize this
+aspect of oneTBB behavior should be able to do it.
+
+This problem can be considered from another angle. Essentially, if the user can indicate
+where parallel computation ends, they can also indicate where it starts.
+
+<img src="parallel_block_introduction.png" width=800>
+
+With this approach, the user not only releases threads when necessary but also specifies a
+programmable block where worker threads should expected new work coming regularly
+to the executing arena.
+
+Let’s add new state to the existing state machine. To represent "Parallel Block" state.
+
+> **_NOTE:_** The "Fast leave" state is colored Grey just for simplicity of the chart.
+              Let's assume that arena was created with the "Delayed leave". 
+              The logic demonstrated below is applicable to the "Fast leave" as well.
+
+<img src="parallel_block_state_initial.png" width=800>
+
+This state diagram leads to several questions. There are some of them:
+* What if there are multiple Parallel Blocks?
+* If “End of Parallel Block” leads back to “Delayed leave” how soon threads
+  will be released from arena?
+  * What if we indicated that threads should leave arena after the "Parallel Block"?
+  * What if we just indicated the end of the "Parallel Block"?
+
+The extended state machine aims to answer these questions.
+* The first call to the “Start of PB” will transition into the “Parallel Block” state.
+* The last call to the “End of PB” will transition back to the “Delayed leave” state
+  or into the "One-time Fast leave" if it is indicated that threads should leave sooner.
+* Concurrent or nested calls to the “Start of PB” or the “End of PB”
+  increment/decrement reference counter.
+
+<img src="parallel_block_state_final.png" width=800>
+
+Let's consider the semantics that an API for explicit parallel blocks can provide:
+* Start of a parallel block:
+  * Indicates the point from which the scheduler can use a hint and keep threads in the arena
+    for longer.
+  * Serves as a warm-up hint to the scheduler:
+    * Allows worker threads to be available by the time real computation starts.
```

Comment:
Do we want the thread warmup to always happen at the start of a block, or do we want to allow users having control over it?
Also, do we promise thread availability "by the time the real computation starts"? I do not think we do, because a) in case too little time has passed after the block start, threads might not yet come, and b) in case too much time has passed, threads might leave.
Maybe more accurate description is like "Allows reducing computation start delays by initiating the wake-up of worker threads in advance".

Reply:
> Do we want the thread warmup to always happen at the start of a block, or do we want to allow users having control over it?

It seems we already discussed this before and agreed that it should be always on because logically start of parallel block should not only indicate the open brace for parallel computation but also start of these computations.